### PR TITLE
Add getParent(Filter)

### PR DESCRIPTION
**Pull request by tdurieux**
This pull request adds the method ```getParent(Filter)``` in order to be able to select a specific parent.
The method ```return null``` if no match is found

### CODE REVIEW

File: src/test/java/spoon/test/parent/ParentTest.java [hunk 1]
```diff
@@ -215,4 +216,61 @@ public void scan(CtElement e) {
...
 
+	@Test
+	public void testGetParentWithFilter() throws Exception {
+		// addType should set Parent
+		CtClass<Foo> clazz = (CtClass<Foo>) factory.Class().getAll().get(0);
+
+		CtMethod<Object> m = clazz.getMethod("m");
+		CtExpression statement = ((CtAssignment)((CtAssignment)m.getBody().getStatement(3)).getAssignment()).getAssignment();
+		CtPackage ctPackage = statement
+				.getParent(new AbstractFilter<CtPackage>(CtPackage.class) {
```
Comment (monperrus) [inline src/test/java/spoon/test/parent/ParentTest.java]: use a TypeFilter?

File: src/test/java/spoon/test/parent/ParentTest.java [hunk 2]
```diff
@@ -215,4 +216,61 @@ public void scan(CtElement e) {
...
+						return true;
+					}
+				});
+		assertEquals(Foo.class.getPackage().getName(), ctPackage.getQualifiedName());
+
+		CtStatement ctStatement = statement
+				.getParent(new AbstractFilter<CtStatement>(CtStatement.class) {
+					@Override
+					public boolean matches(CtStatement element) {
+						return element.getParent() instanceof CtStatementList;
```
Comment (monperrus) [inline src/test/java/spoon/test/parent/ParentTest.java]: call super?

File: src/test/java/spoon/test/parent/ParentTest.java [hunk 3]
```diff
@@ -215,4 +216,61 @@ public void scan(CtElement e) {
...
+					}
+				});
+		assertEquals(m.getBody().getStatement(3), ctStatement);
+
+		m = clazz.getMethod("internalClass");
+		CtStatement ctStatement1 = m.getElements(
+				new AbstractFilter<CtStatement>(CtStatement.class) {
+					@Override
+					public boolean matches(CtStatement element) {
+						return element instanceof CtLocalVariable;
```
Comment (monperrus) [inline src/test/java/spoon/test/parent/ParentTest.java]: super?

File: src/test/java/spoon/test/parent/ParentTest.java [hunk 4]
```diff
@@ -215,4 +216,61 @@ public void scan(CtElement e) {
...
+					public boolean matches(CtType element) {
+						return !element.isAnonymous() && element.isTopLevel();
+					}
+				});
+		assertEquals(clazz, parent);
+		assertNotEquals(ctStatement1.getParent(CtType.class), parent);
+
+		// not present element
+		CtWhile ctWhile = ctStatement1
+				.getParent(new AbstractFilter<CtWhile>(CtWhile.class) {
```
Comment (monperrus) [inline src/test/java/spoon/test/parent/ParentTest.java]: TypeFilter?

File: src/test/java/spoon/test/parent/ParentTest.java [hunk 5]
```diff
@@ -215,4 +216,61 @@ public void scan(CtElement e) {
 
 	}
 
+	@Test
+	public void testGetParentWithFilter() throws Exception {
+		// addType should set Parent
+		CtClass<Foo> clazz = (CtClass<Foo>) factory.Class().getAll().get(0);
+
+		CtMethod<Object> m = clazz.getMethod("m");
+		CtExpression statement = ((CtAssignment)((CtAssignment)m.getBody().getStatement(3)).getAssignment()).getAssignment();
```
Comment (monperrus) [inline src/test/java/spoon/test/parent/ParentTest.java]: add an assert on the toString to make clear what statement is matched

File: src/test/java/spoon/test/parent/ParentTest.java (hunk 3 above)
Comment (tdurieux) [inline src/test/java/spoon/test/parent/ParentTest.java]: No need, the super call returns always true.

File: src/test/java/spoon/test/parent/ParentTest.java (hunk 2 above)
Comment (tdurieux) [inline src/test/java/spoon/test/parent/ParentTest.java]: No need, the super call returns always true.

File: src/test/java/spoon/test/parent/ParentTest.java (hunk 3 above)
Comment (monperrus) [inline src/test/java/spoon/test/parent/ParentTest.java]: It seems it returns "type.isAssignableFrom(element.getClass())"

--Martin

**Comment by monperrus** [comment]
If we change the constructor of "AbstractFilter(Class<?> type)" to "AbstractFilter(Class<? super T> type)" we can now use diamonds (as what was done with TypeFilter). 

Would you add this change in this PR?

**Pull request merged**